### ISSUE TITLE
HTTP digest authentication forced for each page refresh

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2104]
|<img alt="Cyrus Daboo's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| Cyrus Daboo reported|
|-|-|
|Trac ID|trac#2104|
|Type|defect|
|Created|2006-09-18 18:21:16Z|

When using HTTP digest authentication, a client is forced to authenticate to a page even after a successful previous authentication (e.g. when a browser refreshes a page). This occurs because the server discards the 'opaque' parameter in the digest authentication for each successful client login, so thatsubsequent attempt by the client to use the same opaque value fail.

Two solutions for this:

1) Do away with 'opaque' altogether.

2) Cache the 'opaque' value. A good example of this is apache mod_auth_digest which uses a garbage collected map.

Attachments:

* [twisted.web2.auth.digest.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/a8c/a8c97ea909919d6c4c818c1460fb9349eed14423/7590b285e921737cc4eb3e6a1b065432fc8a2c66.patch) (3618 bytes) - added by Cyrus Daboo on 2006-12-07 22:00:38Z - 
* [twisted.web2.test.test_httpauth.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/a8c/a8c97ea909919d6c4c818c1460fb9349eed14423/e4e922d5815de188f9ca80303e4df624c915a37c.patch) (5711 bytes) - added by Cyrus Daboo on 2006-12-07 22:00:54Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__2104 2104
type__defect defect
reporter__Cyrus_Daboo Cyrus Daboo
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web2 web2
keywords__ 
time__1158603676000000 1158603676000000
changetime__1173121027000000 1173121027000000
version__None None
owner__ 
cc__wsanchez cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="Cyrus Daboo's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|Cyrus Daboo commented|
|-|-|

Attached is a patch that does a better job of dealing with client re-authentication.

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_3"></a>|@dreid set status to `assigned`|
|-|-|

branched to [branches/web2-digest-2104](branches/web2-digest-2104)

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_4"></a>|@dreid commented|
|-|-|

This includes the implementation of [#2280](https://github.com/twisted/twisted/issues/2280)

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)|@dreid removed owner<br>@dreid set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to @dreid|
|-|-|

 * None of the error cases for verifyOpaque are tested
 * getChallenge lacks a docstring
 * Thanks for cleaning up that authRequest mess in test_httpauth.py
 * testGetChallenge is missing a docstring
 * testResponse is missing a docstring
 * testMultiResponse is missing a docstring
 * testFailsWithDifferentMethod is missing a docstring
 * testCheckHash imports md5, but md5 is imported already
 * testNoUsername is missing a docstring
 * testMultipleWWWAuthenticateSchemes is missing a docstring
 * testAuthorizationAgainstMultipleSchemes is missing a docstring
 * Of the docstrings present, the formatting is weird.  Can you normalize it to the conventional
```
def foo():
    """
    bar
    """
```
 * In calcHA1, adding preHA1 seems overly complicated.  Is it the crypto function's responsibility to know that preHA1 strings are hex encoded?  What is the behavior if preHA1 is given and the realm, username, password triple is non-None?  What if they produce conflicting results?  What if pszAlg is md5-sess?  Could calcHA1 be left alone and instead, checkHash just decode the hash and pass it in?

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_7"></a>|@dreid commented|
|-|-|

Ok, all the other issues are straightforward so I'm going to address the calcHA1 issues:

* In calcHA1, adding preHA1 seems overly complicated. Is it the crypto function's responsibility to know that preHA1 strings are hex encoded?
 * calcHA1 returns a hex string, because that is the format HA1 is supposed to be in.  A precalculated HA1 is most likely going to be stored as a hex string, and probably resulting from a call to calcHA1.  It seems silly to concern the calling API with converting the HA1 from a hex string.

* What is the behavior if preHA1 is given and the realm, username, password triple is non-None?
 * I think it's sane that if preHA1 is given than any username, realm, password values are ignored.  I will take this as a documentation issue.

* What if they produce conflicting results?
 * It's impossible to verify if the pszUsername, pszRealm, pszPassword passed in would produce conflicting results with the preHA1 passed in.  The point of preHA1 is so that we never have to know the plaintext password.  At any point where we could pass in a pszPassword that wouldn't produce conflicting results, we'd be defeating the purpose of using preHA1.

* What if pszAlg is md5-sess?
 * If the algorithm is md5-sess then calcHA1 adds the nonce, and cnonce to the HA1 before returning it, just like in the case where we lack a precalculated HA1

* Could calcHA1 be left alone and instead, checkHash just decode the hash and pass it in?
 * If you mean call hash.decode('hex'), then yes it could do that, but I think it's kind of silly. (see my response to the first issue raised.)  If however you mean breaking the hash passed to checkHash into username, realm, password and passing those values to calcHA1 then no, it could not be done, the hash passed to checkHash should be an md5 or sha1 hash, so it can not be reversed.

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)|@dreid set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_9"></a>|@glyph set owner to @dreid|
|-|-|

Getting to a complete review of this is going to take a while, but in the meanwhile I think that you should deal with this issue:

 * "I think it's sane that if preHA1 is given than any username, realm, password values are ignored. I will take this as a documentation issue."

I disagree.  Silently ignoring these values sounds like a recipie for disaster, since it will be impossible to tell what's going on without instrumenting the implementation.  Even if they're documented as being ignored, you might not be noticing that you're passing both.  It should most likely fail loudly, since this should never happen.

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_10"></a>|@dreid removed owner|
|-|-|

I've updated the branch to complain very loudly (raise a TypeError with an explanation) when any of pszUsername, pszRealm, or pszPassword are passed in with a preHA1 argument.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_11"></a>|@therve set owner to @dreid|
|-|-|

The branch needs to be merged forward.

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)|@dreid removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_13"></a>|@dreid commented|
|-|-|

Merged forward to [branches/web2-digest-2104-2](branches/web2-digest-2104-2)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_14"></a>|@therve commented|
|-|-|

Quick comments, I don't feel like giving the go on that, but I hope it'll help:
 * md5.md5 and sha.sha are used sometimes, only md5.new and sha.new should be used
 * unused imports in test_httpauth
 * calcHA1 and calcResponse used uppercased variables names, and builtin hash
 * I'd like to see these methods in a common module, but that's a separate issue

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)|@dreid set owner to @dreid<br>@dreid set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_16"></a>|@dreid set owner to @therve<br>@dreid set status to `new`|
|-|-|

Replying to [therve](#note_14):
> Quick comments, I don't feel like giving the go on that, but I hope it'll help:
>  * md5.md5 and sha.sha are used sometimes, only md5.new and sha.new should be used
Fixed.

>  * unused imports in test_httpauth
Fixed.

>  * calcHA1 and calcResponse used uppercased variables names, and builtin hash
This was really done because these are largely (with the exception of preHA1 support) verbatim ports from the reference implementation in RFC2617.  However I did change ```hash``` to ```respHash``` to more closely follow the reference implementation.

>  * I'd like to see these methods in a common module, but that's a separate issue
To be honest, so would I, but my previous attempt at placing the bulk of the digest implementation outside of twisted.web2 failed.  So I'm not really in the mood to press it right now.

In the same vien of not overloading builtins i changed hash in the interface and implementation of checkHash to digestHash.

Thanks for the review, but you left the review keyword and left it unassigned so I didn't know this had been reviewed.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_18"></a>|@therve commented|
|-|-|

Replying to [dreid](#note_16):

> Thanks for the review, but you left the review keyword and left it unassigned so I didn't know this had been reviewed.

Sorry for that, that was a kind of shy review. I'll look at this asap.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_19"></a>|@therve set owner to @dreid|
|-|-|

 * DigestCredentialFactory.decode is missing a docstring.
 * In this method, I'd prefer to see:
```
try:
    username = auth['username']
except AttributeError:
    raise error.LoginFailed('Invalid response, no username given')
```

except if you want to test for empty username ?

 * Always here, I suspect if opaque and nonce aren't in the headers it should fail, so maybe you should test their presence like username
 * DigestCredentialFactory.verifyOpaque should use opaqueParts and keyParts as variable names

That's it for now!

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_20"></a>|@dreid set owner to @therve|
|-|-|

Replying to [therve](#note_19):
>  * DigestCredentialFactory.decode is missing a docstring.
Done

>  * In this method, I'd prefer to see:
> ```
> try:
>     username = auth['username']
> except AttributeError:
>     raise error.LoginFailed('Invalid response, no username given')
> ```

I generally disapprove of this when things like __contains__ and .get are available.

>
> except if you want to test for empty username ?

And yes in fact that is one of the conditions I'm checking for, I've updated the testcase for it.

>  * Always here, I suspect if opaque and nonce aren't in the headers it should fail, so maybe you should test their presence like username

I've added code so that we fail early if opaque and nonce are not in the response, however empty keys will be caught by verifyOpaque (added assertions for this in the appropriate testcases)

>  * DigestCredentialFactory.verifyOpaque should use opaqueParts and keyParts as variable names

Done.

> That's it for now!

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_21"></a>|@therve set owner to @dreid|
|-|-|

Use ```if preHA1 is None``` instead of ```if not preHA1```.

Everything else is really fine. Just be aware that the merge will probably conflict with [#2460](https://github.com/twisted/twisted/issues/2460) (both modifying the test_httpauth).

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_22"></a>|@dreid set status to `closed`|
|-|-|

(In [19763]) Merge web2-digest-2104-2

Author: dreid
Reviewer: therve, glyph, exarkun
Fixes [#2104](https://github.com/twisted/twisted/issues/2104)
Fixes [#2280](https://github.com/twisted/twisted/issues/2280)

Use a verifiable signed opaque to allow clients to reauthenticate 
without forcing a challenge/response authentication to take place 
for every request.

Also make the DigestedCredentials support checking the response against
a precalculated hash of username:realm:password, like that stored in 
.htdigest files.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
